// webdav/dist/interface/dav.js
"use strict";

var path = require("path");
var xml2js = require("xml2js");

function findKey(baseKey, obj) {
    return Object.keys(obj).find(function __findBaseKey(itemKey) {
        var match = /^[a-z0-9]+:(.+)$/i.exec(itemKey);
        return match ? match[1] === baseKey : itemKey === baseKey;
    });
}

function getSingleValue(item) {
    return Array.isArray(item) ? getSingleValue(item[0]) : item;
}

function getValueForKey(key, obj) {
    var keys = void 0,
        i = void 0,
        keyCount = void 0;
    if (typeof obj === "object") {
        var actualKey = findKey(key, obj);
        if (actualKey && typeof obj[actualKey] !== "undefined") {
            return obj[actualKey];
        }
    }
    return undefined;
}

function parseXML(xml) {
    var parser = new xml2js.Parser({ ignoreAttrs: true });
    return new Promise(function (resolve, reject) {
        parser.parseString(xml, function __handleParseResult(err, result) {
            if (err) {
                return reject(err);
            }
            return resolve(result);
        });
    });
}

function propsToStat(props, filename) {
    // Last modified time, raw size, item type and mime
    var lastMod = getSingleValue(getValueForKey("getlastmodified", props));
    var rawSize = getSingleValue(getValueForKey("getcontentlength", props)) || "0";
    var resourceType = getSingleValue(getValueForKey("resourcetype", props));
    var mimeType = getSingleValue(getValueForKey("getcontenttype", props));
    var type = getValueForKey("collection", resourceType) ? "directory" : "file";
    var stat = {
        filename: filename,
        basename: path.basename(filename),
        lastmod: lastMod,
        size: parseInt(rawSize, 10),
        type: type
    };
    if (type === "file") {
        stat.mime = mimeType ? mimeType.split(";")[0] : "";
    }
    return stat;
}

function translateDiskSpace(value) {
    switch (value.toString()) {
        case "-3":
            return "unlimited";
        case "-2":
        /* falls-through */
        case "-1":
            // -1 is non-computed
            return "unknown";
        default:
            return parseInt(value, 10);
    }
}

module.exports = {
    getSingleValue: getSingleValue,
    getValueForKey: getValueForKey,
    parseXML: parseXML,
    propsToStat: propsToStat,
    translateDiskSpace: translateDiskSpace
};
