// webdav/dist/interface/directoryContents.js
"use strict";

var pathPosix = require("path-posix");
var joinURL = require("url-join");

var _require = require("../merge.js"),
    merge = _require.merge;

var responseHandlers = require("../response.js");
var urlTools = require("../url.js");
var davTools = require("./dav.js");
var request = require("../request.js");
var encodePath = request.encodePath;
var fetch = request.fetch;

var getValueForKey = davTools.getValueForKey;
var getSingleValue = davTools.getSingleValue;

function getDirectoryContents(remotePathRaw, options) {
    // Strip the ending slash
    var remotePath = remotePathRaw.replace(/\/$/, "");
    // Join the URL and path for the request
    var fetchURL = joinURL(options.remoteURL, encodePath(remotePath));
    var fetchOptions = {
        method: "PROPFIND",
        headers: merge({
            Depth: 1
        }, options.headers),
        agent: options.agent
    };
    return fetch(fetchURL, fetchOptions).then(responseHandlers.handleResponseCode).then(function __handleResponseFormat(res) {
        // Convert response to text
        return res.text();
    }).then(davTools.parseXML).then(function __handleResult(result) {
        return getDirectoryFiles(result, options.remotePath, remotePath);
    });
}

function getDirectoryFiles(result, serverBasePath, requestPath) {
    var remoteTargetPath = pathPosix.join(serverBasePath, requestPath);
    // Extract the response items (directory contents)
    var multiStatus = getValueForKey("multistatus", result);
    var responseItems = getValueForKey("response", multiStatus);
    return responseItems
    // Filter out the item pointing to the current directory (not needed)
    .filter(function __filterResponseItem(item) {
        var href = getSingleValue(getValueForKey("href", item));
        href = urlTools.normaliseHREF(href);
        href = urlTools.normalisePath(href);
        return href !== serverBasePath && href !== remoteTargetPath;
    })
    // Map all items to a consistent output structure (results)
    .map(function __mapResponseItem(item) {
        // HREF is the file path (in full)
        var href = getSingleValue(getValueForKey("href", item));
        href = urlTools.normaliseHREF(href);
        // Each item should contain a stat object
        var propStat = getSingleValue(getValueForKey("propstat", item));
        var props = getSingleValue(getValueForKey("prop", propStat));
        // Process the true full filename (minus the base server path)
        var filename = serverBasePath === "/" ? urlTools.normalisePath(href) : urlTools.normalisePath(pathPosix.relative(serverBasePath, href));
        return davTools.propsToStat(props, filename);
    });
}

module.exports = {
    getDirectoryContents
};
