// webdav/dist/interface/directoryExists.js
"use strict";

const joinURL = require("url-join");

var _require = require("../response.js");

const handleResponseCode = _require.handleResponseCode;

var _require2 = require("../request.js");

const encodePath = _require2.encodePath,
      prepareRequestOptions = _require2.prepareRequestOptions,
      request = _require2.request;

var _require3 = require("./dav.js");

const getDeepValueForKey = _require3.getDeepValueForKey,
      parseXML = _require3.parseXML;


function directoryExists(dirPath, options) {
    const requestOptions = {
        url: joinURL(options.remoteURL, encodePath(dirPath), "/"),
        method: "PROPFIND",
        headers: {
            Accept: "text/plain",
            Depth: 0
        },
        responseType: "text"
    };
    prepareRequestOptions(requestOptions, options);
    return request(requestOptions).then(handleResponseCode).then(res => res.data).then(parseXML).then(result => {
        console.log(JSON.stringify(result, undefined, 2));
        // const multiStatus = getValueForKey("multistatus", result);
        // const responseItems = getValueForKey("response", multiStatus);
        const isCollection = getDeepValueForKey("multistatus.response.propstat.prop.resourcetype.collection", result);
        console.log(isCollection);
    }).catch(err => {
        if (/status/i.test(err.message) && /\b404\b/.test(err.message)) {
            return false;
        }
        throw err;
    });
}

module.exports = {
    directoryExists
};
