// webdav/dist/interface/getFile.js
"use strict";

var joinURL = require("url-join");
var responseHandlers = require("../response.js");
var request = require("../request.js");
var encodePath = request.encodePath;
var fetch = request.fetch;

function getFileContentsBuffer(filePath, options) {
    return makeFileRequest(filePath, options).then(function (res) {
        return typeof res.buffer === "function" ? res.buffer() : res.arrayBuffer();
    });
}

function getFileContentsString(filePath, options) {
    return makeFileRequest(filePath, options).then(function (res) {
        return res.text();
    });
}

function makeFileRequest(filePath, options) {
    var fetchURL = joinURL(options.remoteURL, encodePath(filePath));
    var fetchOptions = {
        method: "GET",
        headers: options.headers,
        agent: options.agent
    };
    return fetch(fetchURL, fetchOptions).then(responseHandlers.handleResponseCode);
}

function getFileLink(filePath, options) {
    var fetchURL = joinURL(options.remoteURL, encodePath(filePath));
    var protocol = /^https:/i.test(fetchURL) ? "https" : "http";
    if (options.headers.Authorization) {
        if (/^Basic /i.test(options.headers.Authorization) === false) {
            throw new Error("Failed retrieving download link: Invalid authorisation method");
        }
        var authPart = options.headers.Authorization.replace(/^Basic /i, "").trim();
        var authContents = Buffer.from(authPart, "base64").toString("utf8");
        fetchURL = fetchURL.replace(/^https?:\/\//, `${protocol}://${authContents}@`);
    }
    return fetchURL;
}

module.exports = {
    getFileContentsBuffer: getFileContentsBuffer,
    getFileContentsString: getFileContentsString,
    getFileLink: getFileLink
};
