// basic-ftp/dist/FileInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var FileType;
(function (FileType) {
    FileType[FileType["Unknown"] = 0] = "Unknown";
    FileType[FileType["File"] = 1] = "File";
    FileType[FileType["Directory"] = 2] = "Directory";
    FileType[FileType["SymbolicLink"] = 3] = "SymbolicLink";
})(FileType = exports.FileType || (exports.FileType = {}));
/**
 * Describes a file, directory or symbolic link. Note that FTP file listings are not standardized. It depends
 * on the operating system of the FTP server how complete the information is.
 */
class FileInfo {
    constructor(name) {
        this.name = "";
        this.type = FileType.Unknown;
        this.size = 0;
        this.permissions = { user: 0, group: 0, world: 0 };
        this.hardLinkCount = 0;
        this.link = "";
        this.group = "";
        this.user = "";
        this.date = "";
        this.name = name;
    }
    get isDirectory() {
        return this.type === FileType.Directory;
    }
    get isSymbolicLink() {
        return this.type === FileType.SymbolicLink;
    }
    get isFile() {
        return this.type === FileType.File;
    }
}
exports.FileInfo = FileInfo;
FileInfo.Permission = {
    Read: 4,
    Write: 2,
    Execute: 1
};
