// homey-api/lib/HomeyAPI/HomeyAPIV3/Item.js
'use strict';

const EventEmitter = require('../../EventEmitter');
const RealtimeConsumer = require('./RealtimeConsumer');

/**
 * A superclass for all CRUD Items.
 * @class
 * @hideconstructor
 * @memberof HomeyAPIV3
 */
class Item extends EventEmitter {
  static ID = null; // Set by Manager.js

  /**
   * The ID of the Item.
   * @type {string}
   */
  id;
  /**
   * The URI of the Item, e.g. `homey:foo:bar`.
   * @type {string}
   */
  uri;

  constructor({ id, homey, manager, properties }) {
    super();

    Object.defineProperty(this, 'id', {
      value: id,
      enumerable: true,
      writable: false,
    });

    Object.defineProperty(this, 'uri', {
      value: `homey:${this.constructor.ID}:${this.id}`,
      enumerable: true,
      writable: false,
    });

    Object.defineProperty(this, '__homey', {
      value: homey,
      enumerable: false,
      writable: false,
    });

    Object.defineProperty(this, '__manager', {
      value: manager,
      enumerable: false,
      writable: false,
    });

    Object.defineProperty(this, '__connected', {
      value: false,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__lastUpdated', {
      value: new Date(),
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__realtimeConsumer', {
      value: new RealtimeConsumer({
        subscribe: (uri, handlers) => this.homey.subscribe(uri, handlers),
        getUri: () => this.uri,
        debug: (...props) => this.__debug(...props),
        setConnected: (connected) => {
          this.__connected = connected;
        },
        onConnect: () => {
          this.onConnect();
        },
        onDisconnect: () => {
          this.onDisconnect();
        },
        onReconnect: () => {
          this.onReconnect();
        },
        onEvent: (event, data) => {
          if (event === 'update') {
            this.__update(this.constructor.transformGet({ ...data }));
            return;
          }

          if (event === 'delete') {
            this.__delete();
            return;
          }

          this.emit(event, data);
        },
      }),
      enumerable: false,
      writable: false,
    });

    // Set Properties
    for (const [key, value] of Object.entries(properties)) {
      if (key === 'id') continue;

      Object.defineProperty(this, key, {
        value,
        enumerable: true,
        writable: true,
      });
    }
  }

  /**
   * The Manager of the Item.
   * @type {HomeyAPIV3.Manager}
   */
  get manager() {
    return this.__manager;
  }

  /**
   * The Homey of the Item.
   * @type {HomeyAPIV3}
   */
  get homey() {
    return this.__homey;
  }

  __debug(...props) {
    this.manager.__debug(`[${this.constructor.name}:${this.id}]`, ...props);
  }

  __update(properties, { emitEvent = true } = {}) {
    for (const [key, value] of Object.entries(properties)) {
      if (key === 'id') continue;
      this[key] = value;
    }

    this.__lastUpdated = new Date();

    if (emitEvent) {
      this.emit('update', properties);
    }
  }

  __delete() {
    this.emit('delete');
    this.destroy();
  }

  /**
   * Connect to this item's Socket.io namespace.
   */
  async connect() {
    await this.__realtimeConsumer.connect();
  }

  /**
   * Discconnect from this item's Socket.io namespace.
   */
  async disconnect() {
    await this.__realtimeConsumer.disconnect();
  }

  onConnect() {
    // Overload Me
  }

  onReconnect() {
    // Overload Me
  }

  onDisconnect() {
    // Overload Me
  }

  destroy() {
    // Remove all event listeners
    this.removeAllListeners();

    // Disconnect from Socket.io
    this.disconnect().catch(() => { });
  }

  static transformGet(item) {
    return item;
  }

  static transformSet(item) {
    return item;
  }
}

module.exports = Item;
