// homey-api/lib/HomeyAPI/HomeyAPIV3.js
'use strict';

const Util = require('../Util');
const HomeyAPI = require('./HomeyAPI');
const HomeyAPIError = require('./HomeyAPIError');
const ManagerApps = require('./HomeyAPIV3/ManagerApps');
const ManagerDrivers = require('./HomeyAPIV3/ManagerDrivers');
const ManagerDevices = require('./HomeyAPIV3/ManagerDevices');
const ManagerFlow = require('./HomeyAPIV3/ManagerFlow');
const ManagerFlowToken = require('./HomeyAPIV3/ManagerFlowToken');
const ManagerInsights = require('./HomeyAPIV3/ManagerInsights');
const ManagerUsers = require('./HomeyAPIV3/ManagerUsers');
const ManagerZones = require('./HomeyAPIV3/ManagerZones');
const DiscoveryManager = require('./HomeyAPIV3/DiscoveryManager');
const Manager = require('./HomeyAPIV3/Manager');
const SocketSession = require('./HomeyAPIV3/SocketSession');
const SubscriptionRegistry = require('./HomeyAPIV3/SubscriptionRegistry');

/**
 * An authenticated Homey API. Do not construct this class manually.
 * @class
 * @hideconstructor
 * @extends HomeyAPI
 */
class HomeyAPIV3 extends HomeyAPI {
  static MANAGERS = {
    ManagerApps,
    ManagerDrivers,
    ManagerDevices,
    ManagerFlow,
    ManagerFlowToken,
    ManagerInsights,
    ManagerUsers,
    ManagerZones,
  };

  static DEFAULT_TIMEOUT = 1000 * 10;
  static HANDSHAKE_TIMEOUT = 1000 * 10;
  static SUBSCRIBE_TIMEOUT = 1000 * 10;

  constructor({
    properties,
    strategy = [
      HomeyAPI.DISCOVERY_STRATEGIES.MDNS,
      HomeyAPI.DISCOVERY_STRATEGIES.CLOUD,
      HomeyAPI.DISCOVERY_STRATEGIES.LOCAL,
      HomeyAPI.DISCOVERY_STRATEGIES.LOCAL_SECURE,
      HomeyAPI.DISCOVERY_STRATEGIES.REMOTE_FORWARDED,
    ],
    baseUrl = null,
    token = null,
    session = null,
    reconnect = true,
    api,
    debug,
  }) {
    super({ properties, api, debug });

    this.__refreshMap = {};

    Object.defineProperty(this, '__baseUrl', {
      value: baseUrl,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__strategyId', {
      value: null,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__reconnect', {
      value: reconnect,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__destroyed', {
      value: false,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__token', {
      value: token,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__session', {
      value: session,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__strategies', {
      value: Array.isArray(strategy) ? strategy : [strategy],
      enumerable: false,
      writable: false,
    });

    Object.defineProperty(this, '__managers', {
      value: {},
      enumerable: false,
      writable: false,
    });

    Object.defineProperty(this, '__baseUrlPromise', {
      value: typeof baseUrl === 'string' ? Promise.resolve(baseUrl) : null,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__loginPromise', {
      value: null,
      enumerable: false,
      writable: true,
    });

    this.__discoveryManager = new DiscoveryManager(this);
    this.__socketSession = new SocketSession(this);
    this.__subscriptionRegistry = new SubscriptionRegistry({
      homey: this,
      session: this.__socketSession,
    });

    this.__bindSocketSessionEvents();

    this.generateManagersFromSpecification();
  }

  /*
   * Get the Homey's base URL promise
   */
  get baseUrl() {
    return (async () => {
      if (!this.__baseUrlPromise) {
        this.__baseUrlPromise = (async () => {
          const { baseUrl } = await this.discoverBaseUrl();
          return baseUrl;
        })();
        this.__baseUrlPromise.catch(() => {});
      }

      return this.__baseUrlPromise;
    })();
  }

  get strategyId() {
    return this.__strategyId;
  }

  get __socket() {
    return this.__socketSession.socket;
  }

  get __homeySocket() {
    return this.__socketSession.homeySocket;
  }

  __bindSocketSessionEvents() {
    for (const event of [
      'connect',
      'disconnect',
      'reconnect',
      'reconnect_attempt',
      'reconnecting',
      'reconnect_error',
      'connect_error',
      'error',
    ]) {
      this.__socketSession.on(event, (...args) => {
        this.emit(event, ...args);
      });
    }
  }

  /*
   * Generate Managers from JSON specification
   * A manager instance is created when it's first accessed
   */

  getSpecification() {
    return require('../../assets/specifications/HomeyAPIV2.json');
  }

  generateManagersFromSpecification() {
    const { managers } = this.getSpecification();
    Object.entries(managers).forEach(([managerName, manager]) => {
      this.generateManagerFromSpecification(managerName, manager);
    });
  }

  generateManagerFromSpecification(managerName, manager) {
    Object.defineProperty(this, manager.idCamelCase, {
      get: () => {
        if (!this.__managers[managerName]) {
          const ManagerClass = this.constructor.MANAGERS[managerName]
            ? this.constructor.MANAGERS[managerName]
            : (() => {
                return class extends Manager {};
              })();

          ManagerClass.ID = manager.id;

          this.__managers[managerName] = new ManagerClass({
            homey: this,
            items: manager.items || {},
            operations: manager.operations || {},
          });
        }

        return this.__managers[managerName];
      },
      enumerable: false,
    });
  }

  /*
   * Discover the URL to talk to Homey
   * We prefer localSecure, because it's fastest and most secure
   * If that doesn't work, we prefer local OR mdns, whichever is fastest
   * Finally, we fallback to cloud
   */

  async discoverBaseUrl() {
    const discoveryResult = await this.__discoveryManager.discoverBaseUrl();
    const result = this.__applyDiscoveryResult(discoveryResult);
    return result;
  }

  __applyDiscoveryResult({ baseUrl, strategyId, homeyVersion, homeyTier }) {
    this.__baseUrl = baseUrl;
    this.__baseUrlPromise = Promise.resolve(baseUrl);
    this.__strategyId = strategyId;
    this.__tierCache[this.id] = homeyTier;

    if (homeyVersion != null) {
      this.version = homeyVersion;
    }

    return {
      baseUrl,
      strategyId,
      homeyVersion,
      homeyTier,
    };
  }

  async call({
    $timeout = this.constructor.DEFAULT_TIMEOUT,
    method,
    headers,
    path,
    body,
    json = true,
    isRetryAfterRefresh = false,
    shouldRetry = true,
  }) {
    const token = this.__token;
    const baseUrl = await this.baseUrl;

    method = String(method).toUpperCase();

    headers = {
      ...headers,
      'X-Homey-ID': this.id,
    };

    if (token && path !== '/api/manager/users/login') {
      headers['Authorization'] = `Bearer ${token}`;
    }

    const originalBody = body;

    if (['PUT', 'POST'].includes(method)) {
      if (body && json === true) {
        headers['Content-Type'] = 'application/json';
        body = JSON.stringify(body);
      }
    } else {
      body = undefined;
    }

    this.__debug(method, `${baseUrl}${path}`);

    const res = await Util.fetch(
      `${baseUrl}${path}`,
      {
        method,
        headers,
        body,
      },
      $timeout,
    );

    const resStatusCode = res.status;
    if (resStatusCode === 204) return undefined;

    const resStatusText = res.status;
    const resHeadersContentType = res.headers.get('Content-Type');
    const version = res.headers.get('x-homey-version');
    if (version) {
      this.__versionCache[this.id] = version;
    }
    const tier = res.headers.get('x-homey-tier');
    if (tier) {
      this.__tierCache[this.id] = tier;
    }

    const resBodyText = await res.text();
    let resBodyJson;
    if (resHeadersContentType && resHeadersContentType.startsWith('application/json')) {
      try {
        resBodyJson = JSON.parse(resBodyText);
        // eslint-disable-next-line no-empty
      } catch (err) {}
    }

    if (!res.ok) {
      // If Session Expired, clear the stored token. Only try if AthomCloudAPI is available so we
      // never try to refresh on local api's.
      if (resStatusCode === 401 && shouldRetry === true && token != null && this.__api != null) {
        this.__debug('Session expired');
        await this.refreshForToken(token, isRetryAfterRefresh);

        if (!isRetryAfterRefresh) {
          return this.call({
            method,
            headers,
            path,
            body: originalBody,
            isRetryAfterRefresh: true,
          });
        }
      }

      if (resBodyJson) {
        throw new HomeyAPIError(
          {
            error: resBodyJson.error,
            error_description: resBodyJson.error_description,
            stack: resBodyJson.stack,
          },
          resStatusCode,
        );
      }

      if (resBodyText) {
        throw new HomeyAPIError(
          {
            error: resBodyText,
          },
          resStatusCode,
        );
      }

      throw new HomeyAPIError(
        {
          error: resStatusText,
        },
        resStatusCode,
      );
    }

    if (typeof resBodyJson !== 'undefined') {
      return resBodyJson;
    }

    return resBodyText;
  }

  async login() {
    if (!this.__loginPromise) {
      this.__loginPromise = Promise.resolve().then(async () => {
        // Check store for a valid Homey.Session
        const store = await this.__getStore();

        if (store && store.token && store.session) {
          this.__debug('Got token from store');
          this.__token = store.token;
          this.__session = store.session;

          await this.baseUrl; // Ensure discovery has completed
          return;
        }

        await this.__refreshSession();
      });

      this.__loginPromise
        .then(() => {
          this.__loginPromise = null;
        })
        .catch((err) => {
          this.__debug('Error Logging In:', err);
          this.__loginPromise = null;
          this.__token = null;
          this.__session = null;
        });
    }

    return this.__loginPromise;
  }

  async __refreshSession() {
    this.__debug('refreshSession');

    if (this.__api) {
      // Create a Session by generating a JWT token on AthomCloudAPI,
      // and then sending the JWT token to Homey.
      const jwtToken = await this.__api.createDelegationToken({ audience: 'homey' });
      const token = await this.users.login({
        $socket: false,
        token: jwtToken,
        shouldRetry: false,
      });
      this.__token = token;

      const session = await this.sessions.getSessionMe({
        $socket: false,
        shouldRetry: false,
      });
      this.__session = session;

      await this.__setStore({ session, token });

      this.__debug('Got token');
      return;
    }

    throw new Error('Cannot Sign In: Missing AthomCloudAPI');
  }

  async refreshSession() {
    if (!this.__refreshSessionPromise) {
      this.__refreshSessionPromise = this.__refreshSession();

      this.__refreshSessionPromise
        .catch((err) => {
          this.__debug('Error refreshing session:', err);
        })
        .finally(() => {
          this.__refreshSessionPromise = null;
        });
    }

    return this.__refreshSessionPromise;
  }

  async logout() {
    this.__token = null;
    this.__session = null;

    await this.__setStore({
      token: null,
      session: null,
    });
  }

  async refreshForToken(token, isRetryAfterRefresh = false) {
    if (this.__token === token && this.__token !== null && this.__refreshMap[token] == null) {
      this.__refreshMap[token] = Promise.resolve().then(async () => {
        await this.__setStore({
          token: null,
          session: null,
        });

        if (!isRetryAfterRefresh) {
          this.__debug('Refreshing token...');
          // If the login fails, the tokens are cleared on the instance. We don't call logout and clear
          // the token on the instance because that could cause a handshake client to fire with a null
          // token. Handshake client also needs to attempt to refresh the token, and if the refresh was
          // already started, it should reuse this promise. That also goes the other way around.
          await this.login();
        }
      });

      this.__refreshMap[token]
        .then(() => {})
        .catch((err) => {
          this.__debug('Error Refreshing Token:', err);
        })
        .finally(() => {
          // Delete after 30 seconds some requests might still be pending and they should be able
          // to receive a rejected promise for this token.
          const cleanupTimer = setTimeout(() => {
            delete this.__refreshMap[token];
            delete this.__refreshMap[token + 'timeout'];
          }, 30 * 1000);

          if (typeof cleanupTimer.unref === 'function') {
            cleanupTimer.unref();
          }

          this.__refreshMap[token + 'timeout'] = cleanupTimer;
        });
    }

    await this.__refreshMap[token];
  }

  /**
   * If Homey is connected to Socket.io.
   * @returns {Boolean}
   */
  isConnected() {
    return this.__socketSession.isConnected();
  }

  async subscribe(uri, handlers) {
    return this.__subscriptionRegistry.subscribe(uri, handlers);
  }

  async __apiRequest({ uri, operation, args, timeout = this.constructor.DEFAULT_TIMEOUT }) {
    return this.__socketSession.requestApi({
      uri,
      operation,
      args,
      timeout,
    });
  }

  async connect() {
    const readyState = await this.__socketSession.ensureConnected();
    await this.__subscriptionRegistry.ensureSessionSubscriptions(readyState);
  }

  isConnecting() {
    return this.__socketSession.isConnecting();
  }

  async disconnect() {
    await this.__socketSession.disconnect();
  }

  destroy() {
    this.__destroyed = true;

    for (const [key, value] of Object.entries(this.__refreshMap)) {
      if (key.endsWith('timeout')) {
        clearTimeout(value);
      }
    }

    this.__refreshMap = {};
    this.__subscriptionRegistry.destroy();
    this.__socketSession.destroy();

    this.removeAllListeners();
  }

  isDestroyed() {
    return this.__destroyed;
  }
  hasScope(scope) {
    if (this.__session == null) {
      this.__debug('Tried to call hasScope without a session present.');
      return true;
    }

    return this.__session.intersectedScopes.some((availableScope) => {
      return this.constructor.instanceOfScope(scope, availableScope);
    });
  }

  static instanceOfScope(scopeOne, scopeTwo) {
    const partsOne = scopeOne.split(':');
    const partsTwo = scopeTwo.split(':');

    let suffixIsEqual = true;

    if (partsOne[1]) {
      suffixIsEqual = partsOne[1] === partsTwo[1];
    }

    return scopeOne.indexOf(partsTwo[0]) === 0 && suffixIsEqual;
  }
}

module.exports = HomeyAPIV3;
